File: Exercises/Java/diet_dual.java
/* This example formulates and solves the dual of the diet model:

     maximize   max v=21pi + 12pc
     subject to  2pi + 0pc <= 20
                 0pi + 1pc <= 10
                 3pi + 2pc <= 31
                 1pi + 2pc <= 11
                 2pi + 1pc <= 12                 
*/

import gurobi.*;

public class diet_dual {
  public static void main(String[] args) {
    try {
      GRBEnv    env   = new GRBEnv("diet_dual.log");
      GRBModel  model = new GRBModel(env);

      // Create variables

    GRBVar pi = model.addVar(0.0, GRB.INFINITY, 0.0, GRB.CONTINUOUS, "pi");
    GRBVar pc = model.addVar(0.0, GRB.INFINITY, 0.0, GRB.CONTINUOUS, "pc");

      // Integrate new variables

      model.update();

      // Set objective: maximize   max v=21pi + 12pc

      GRBLinExpr expr = new GRBLinExpr();
      expr.addTerm(21.0, pi); 
      expr.addTerm(12.0, pc); 
      model.setObjective(expr, GRB.MAXIMIZE);

      // Add constraint: 2pi + 0pc <= 20

      expr = new GRBLinExpr();
      expr.addTerm(2.0, pi); 
      expr.addTerm(0.0, pc); 
      GRBConstr c1 = model.addConstr(expr, GRB.LESS_EQUAL, 20.0, "c1");

      // Add constraint: 0pi + 1pc <= 10

      expr = new GRBLinExpr();
      expr.addTerm(0.0, pi); 
      expr.addTerm(1.0, pc); 
      GRBConstr c2 = model.addConstr(expr, GRB.LESS_EQUAL, 10.0, "c2");

      // Add constraint: 3pi + 2pc <= 31

      expr = new GRBLinExpr();
      expr.addTerm(3.0, pi); 
      expr.addTerm(2.0, pc); 
      GRBConstr c3 = model.addConstr(expr, GRB.LESS_EQUAL, 31.0, "c3");

      // Add constraint: 1pi + 2pc <= 11

      expr = new GRBLinExpr();
      expr.addTerm(1.0, pi); 
      expr.addTerm(2.0, pc); 
      GRBConstr c4 = model.addConstr(expr, GRB.LESS_EQUAL, 11.0, "c4");

      // Add constraint: 2pi + 1pc <= 12

      expr = new GRBLinExpr();
      expr.addTerm(2.0, pi); 
      expr.addTerm(1.0, pc); 
      GRBConstr c5 = model.addConstr(expr, GRB.LESS_EQUAL, 12.0, "c5");

      // Optimize model

      model.optimize();

	  System.out.println("\nVariables:\nV Name" + "\t" + "Value" + "\t" + "Red. Cost");
      System.out.println(pi.get(GRB.StringAttr.VarName)
                         + "\t" + pi.get(GRB.DoubleAttr.X)
                         + "\t" + pi.get(GRB.DoubleAttr.RC));
      System.out.println(pc.get(GRB.StringAttr.VarName)
                         + "\t" + pc.get(GRB.DoubleAttr.X)
                         + "\t" + pc.get(GRB.DoubleAttr.RC));

      System.out.println("\nObj: " + model.get(GRB.DoubleAttr.ObjVal));

	  System.out.println("\nConstraints:\nC Name" + "\t" + "Slack" + "\t" + "Dual Val");
      System.out.println(c1.get(GRB.StringAttr.ConstrName)
                         + "\t" + c1.get(GRB.DoubleAttr.Slack)
                         + "\t" + c1.get(GRB.DoubleAttr.Pi));
      System.out.println(c2.get(GRB.StringAttr.ConstrName)
              + "\t" + c2.get(GRB.DoubleAttr.Slack)
              + "\t" + c2.get(GRB.DoubleAttr.Pi));
      System.out.println(c3.get(GRB.StringAttr.ConstrName)
              + "\t" + c3.get(GRB.DoubleAttr.Slack)
              + "\t" + c3.get(GRB.DoubleAttr.Pi));
      System.out.println(c4.get(GRB.StringAttr.ConstrName)
              + "\t" + c4.get(GRB.DoubleAttr.Slack)
              + "\t" + c4.get(GRB.DoubleAttr.Pi));
      System.out.println(c5.get(GRB.StringAttr.ConstrName)
              + "\t" + c5.get(GRB.DoubleAttr.Slack)
              + "\t" + c5.get(GRB.DoubleAttr.Pi));
                         
      // Dispose of model and environment

      model.dispose();
      env.dispose();

    } catch (GRBException e) {
      System.out.println("Error code: " + e.getErrorCode() + ". " +
                         e.getMessage());
    }
  }
}

File: Exercises/C/diet_dual.c
/* This example formulates and solves the dual of the diet model:

     maximize   max v=21pi + 12pc
     subject to  2pi + 0pc <= 20
                 0pi + 1pc <= 10
                 3pi + 2pc <= 31
                 1pi + 2pc <= 11
                 2pi + 1pc <= 12
*/

#include <stdlib.h>
#include <stdio.h>
#include "gurobi_c.h"

int
main(int   argc,
     char *argv[])
{
  GRBenv   *env   = NULL;
  GRBmodel *model = NULL;
  int 		i;
  int       error = 0;
  char     *varnames[2] = { "pi", "pc" };
  char     *name[2];
  double    sol[2];
  double    rc[2];
  char     *conname[5];
  double    slack[5];
  double    pi[5];
  int       ind[2];
  double    val[2];
  double    obj[2];
  char      vtype[2];
  int       optimstatus;
  double    objval;

  /* Create environment */

  error = GRBloadenv(&env, "diet.log");
  if (error) goto QUIT;

  /* Create an empty model */

  error = GRBnewmodel(env, &model, "diet", 0, NULL, NULL, NULL, NULL, NULL);
  if (error) goto QUIT;


  /* Add variables */

  obj[0] = 21; obj[1] = 12;
  for ( i = 0; i < 2; i++) {
	  vtype[i] = GRB_CONTINUOUS;
  }
  error = GRBaddvars(model, 2, 0, NULL, NULL, NULL, obj, NULL, NULL, vtype, varnames);
  if (error) goto QUIT;

  /* Change objective sense to minimization */

  error = GRBsetintattr(model, GRB_INT_ATTR_MODELSENSE, GRB_MAXIMIZE);
  if (error) goto QUIT;

  /* Integrate new variables */

  error = GRBupdatemodel(model);
  if (error) goto QUIT;


  /* First constraint: 2pi + 0pc <= 20 */

  ind[0] = 0; ind[1] = 1;
  val[0] = 2; val[1] = 0;

  error = GRBaddconstr(model, 2, ind, val, GRB_LESS_EQUAL, 20.0, "c1");
  if (error) goto QUIT;

  /* 2nd constraint: 0pi + 1pc <= 10 */

  ind[0] = 0; ind[1] = 1;
  val[0] = 0; val[1] = 1;

  error = GRBaddconstr(model, 2, ind, val, GRB_LESS_EQUAL, 10.0, "c2");
  if (error) goto QUIT;

  /* 3rd constraint: 3pi + 2pc <= 31 */

  ind[0] = 0; ind[1] = 1;
  val[0] = 3; val[1] = 2;

  error = GRBaddconstr(model, 2, ind, val, GRB_LESS_EQUAL, 31.0, "c3");
  if (error) goto QUIT;

  /* 4th constraint: 1pi + 2pc <= 11 */

  ind[0] = 0; ind[1] = 1;
  val[0] = 1; val[1] = 2;

  error = GRBaddconstr(model, 2, ind, val, GRB_LESS_EQUAL, 11.0, "c4");
  if (error) goto QUIT;

  /* 5th constraint: 2pi + 1pc <= 12 */

  ind[0] = 0; ind[1] = 1;
  val[0] = 2; val[1] = 1;

  error = GRBaddconstr(model, 2, ind, val, GRB_LESS_EQUAL, 12.0, "c5");
  if (error) goto QUIT;


  /* Optimize model */

  error = GRBoptimize(model);
  if (error) goto QUIT;

  /* Write model to 'mip1.lp' */

  error = GRBwrite(model, "diet.lp");
  if (error) goto QUIT;

  /* Capture solution information */

  error = GRBgetintattr(model, GRB_INT_ATTR_STATUS, &optimstatus);
  if (error) goto QUIT;

  error = GRBgetdblattr(model, GRB_DBL_ATTR_OBJVAL, &objval);
  if (error) goto QUIT;

  error = GRBgetstrattrarray(model, GRB_STR_ATTR_VARNAME, 0, 2, name);
  if (error) goto QUIT;

  error = GRBgetdblattrarray(model, GRB_DBL_ATTR_X, 0, 2, sol);
  if (error) goto QUIT;

  error = GRBgetdblattrarray(model, GRB_DBL_ATTR_RC, 0, 2, rc);
  if (error) goto QUIT;

  error = GRBgetstrattrarray(model, GRB_STR_ATTR_CONSTRNAME, 0, 5, conname);
  if (error) goto QUIT;

  error = GRBgetdblattrarray(model, GRB_DBL_ATTR_SLACK, 0, 5, slack);
  if (error) goto QUIT;

  error = GRBgetdblattrarray(model, GRB_DBL_ATTR_PI, 0, 5, pi);
  if (error) goto QUIT;

  printf("\nOptimization complete\n");
  if (optimstatus == GRB_OPTIMAL) {
	printf("\nVariables:\nV Name      Value    Red. Cost\n");
	for (i=0; i<2; i++) {
        printf("%4s       %5.1f     %8.4f\n",name[i],sol[i],rc[i]);
	}

    printf("\nOptimal objective: %.4e\n", objval);

	printf("\nConstraints:\n C Name     Slack    Dual Value\n");
	for (i=0; i<5; i++) {
        printf("%7s    %5.1f     %8.4f\n",conname[i],slack[i],pi[i]);
	}
  } else if (optimstatus == GRB_INF_OR_UNBD) {
    printf("Model is infeasible or unbounded\n");
  } else {
    printf("Optimization was stopped early\n");
  }

QUIT:

  /* Error reporting */

  if (error) {
    printf("ERROR: %s\n", GRBgeterrormsg(env));
    exit(1);
  }

  /* Free model */

  GRBfreemodel(model);

  /* Free environment */

  GRBfreeenv(env);

  return 0;
}
